Benchmark JSON parsing

rossabaker opened the pull request:
Revives the old JSON parser from history and creates a JMH for it to evaluate performance.  They can be run like this:

```
bench/jmh:run -wi 10 -i 10 -f1 -t8
```

rossabaker commented:
I am unsure that I am running the parsers correctly.  The "hot" parser is failing with multiple threads.  Quick commands to demonstrate this (`-t` is the thread count, and iterations are reduced):

This works:

```
bench/jmh:run -wi 0 -i 1 -f1 -t1 .*Countries.*Hot
```

This fails:

```
bench/jmh:run -wi 0 -i 1 -f1 -t2 .*Countries.*Hot
```

J-mie6 commented:
Hmm, that's strange, you're using `runParserThreadSafe` which is what I expect... I've used parsers multithreaded via this mechanism before (but not with JMH) so I'll have to investigate it today.

J-mie6 commented on bench/src/main/scala/parsley/bench/JsonBench.scala:
```diff
@@ -0,0 +1,51 @@
+  def json: Parsley[JValue] = {
+    val jsontoks = LanguageDef("", "", "", false, NotRequired, NotRequired, NotRequired, NotRequired, Set.empty, Set.empty, true, Predicate(Char.isWhitespace))
+    val tok = new TokenParser(jsontoks)
+    lazy val obj: Parsley[JValue] = tok.braces(tok.commaSep(+(tok.stringLiteral <~> tok.colon *> value)).map(pairs => JObject.fromSeq(pairs)))
```
By the way, since this parser was originally written, `+` is pointless. I should have deprecated it really (or removed it outright). The optimisation that it is supposed to perform is now down automatically by the Parsley compiler.

J-mie6 commented on bench/src/main/scala/parsley/bench/JsonBench.scala:
```diff
@@ -0,0 +1,51 @@
+    lazy val array: Parsley[JValue] = tok.brackets(tok.commaSep(value)).map(list => JArray.fromSeq(list))
+    lazy val value: Parsley[JValue] =
+      (tok.stringLiteral.map(JString.apply)
+        <|> tok.symbol("true") *> Parsley.pure(JTrue)
```
Not that it makes a difference (Parsley will optimise it exactly the same) but a more concise way of writing `*> pure` is `#>`

J-mie6 commented on bench/src/main/scala/parsley/bench/JsonBench.scala:
```diff
@@ -0,0 +1,51 @@
+    Source.fromResource(name, getClass.getClassLoader).getLines().mkString("\n")
+
+  def json: Parsley[JValue] = {
+    val jsontoks = LanguageDef("", "", "", false, NotRequired, NotRequired, NotRequired, NotRequired, Set.empty, Set.empty, true, Predicate(Char.isWhitespace))
```
Any reason you didn't use `val jsontoks = LanguageDef.plain.copy(space = Predicate(Char.isWhitespace))`?

J-mie6 replied:
Perhaps this is an even older version of the benchmark parser than the one on `brexit`...

rossabaker replied:
I exhumed this from a GitHub search and didn't find the head of the branch it lived on.  Yes, this is much nicer.

J-mie6 commented:
> Hmm, that's strange, you're using `runParserThreadSafe` which is what I expect... I've used parsers multithreaded via this mechanism before (but not with JMH) so I'll have to investigate it today.

Actually, it appears that there isn't much of a difference between `runParser` and `runParserThreadSafe`, not sure why I made a distinction between them. Just as long as you aren't using `runParserFastUnsafe` in a multi-threaded environment.

J-mie6 commented:
So I've been trying to replicate the bug in scala test (not on your branch), and it's thrown up nothing:

```
"parsers" should "be thread safe when ran correctly" in {
        import scala.concurrent.{Future, ExecutionContext, Await, duration}, ExecutionContext.global, duration._
        implicit val ec = global
        val p = many('a' *> many('b' *> 'b' *> 'b'))
        val as = 1000
        val bs = 2000
        val input = ("a" + "b" * (bs * 3)) * as
        val output = Success(List.fill[Char](as, bs)('b'))

        val futs = Future.traverse(1 to 16)(_ => Future(runParser(p, input)))
        for (out <- Await.result(futs, 20.second)) {
            out should be (output)
        }
    }
```

As far as I'm concerned, this test should be picking up any state copying errors, and it takes a good 2 seconds to run. I wonder if it's a problem with JMH itself interacting poorly with Parsley?

J-mie6 commented on bench/src/main/scala/parsley/bench/JsonBench.scala:
```diff
@@ -0,0 +1,51 @@
+    runParserThreadSafe(json, text).toOption.get
+
+  // Stable instance to warm up
+  val hotJson: Parsley[JValue] = json
```
`hotJson` isn't going to work as you expect by the way, you need to use `.force()` to get it to properly compile the parser in advance, the generation of the instructions is lazy.

J-mie6 commented:
Oh also, there was a set of JMH benchmarks made by Tom Lin here: [https://github.com/tom91136/scala-parser-benchmarks](scala parser benchmarks). I think he ran them single-core, however.

rossabaker commented:
I had been wondering what the difference between `runParser` and `runParserThreadSafe` was.  I switched to the former, and see the same results.  I also changed to throw the parse error instead of an uninformative `None.get`, and can confirm that the error is non-deterministic.

I'll try moving the hot parser out to a singleton, to make sure there's nothing weird having it as an instance variable in the JMH framework.  (I think the `@Scope` annotation I used is right, but maybe not?)

J-mie6 commented:
I'm having a quick check on your generated parser by hand to ensure that there isn't anything shady going on with the states under the hood

rossabaker commented:
Moving the parsers to the companion didn't help.  Lunch is about over, but I'll try something like your unit test with the JSON parser.

rossabaker commented:
Got pulled away, but this also fails, as just a regular app:

```scala

object Concurrent {
  import scala.concurrent._
  import scala.concurrent.duration._
  import scala.concurrent.ExecutionContext.Implicits.global

  val text: String =
    Source.fromResource("countries.geo.json", getClass.getClassLoader).getLines().mkString("\n")

  def parsleyParseHotConcurrent(): Seq[JValue] = {
    val futs: Future[Seq[JValue]] = Future.traverse(1 to 16) { _ =>
      Future(runParser(JsonBench.hotJson, text)).flatMap {
        case Success(x) => Future.successful(x)
        case Failure(e) => Future.failed(new Exception(e.toString))
      }
    }
    Await.result(futs, 20.seconds)
  }

  def main(args: Array[String]) = {
    parsleyParseHotConcurrent()
    ()
  }
}
```

J-mie6 commented:
Ok great, I'll take a look at it tonight + tomorrow evening!

J-mie6 commented:
I've dug myself into a huge hole at the moment: it's possible that when I've dug myself out of it, the bug _might_ be gone, but who knows. It turned out that there was a slight laziness issue with let detection, but fixing it is requiring some serious gutting. Hopefully it won't be too long until I'm in a position to try out your test case and see what's going wrong with it!

J-mie6 commented:
Right, there is possibly a chance that the head of 2.13 fixes the race condition, but I'm going to grab your test case and take a look to see if it fails for me now!

rossabaker commented:
If you want a version of the above test that compiles, https://github.com/http4s/Parsley/tree/non-jmh.  I just sloppily commented out the JMH dependencies, but can be run with `sbt bench/run`.  That branch does not merge your fix, though.

J-mie6 commented:
Ok I can reproduce your issue fine and it's still happening on head. I'll see what I can find out about it today!

J-mie6 commented:
Ok I think I know what it is. I was _very_ lucky actually. The third time I ran the test I got this error:

```
[error] (run-main-0) java.lang.ArrayIndexOutOfBoundsException: Index 37 out of bounds for length 32
[error] java.lang.ArrayIndexOutOfBoundsException: Index 37 out of bounds for length 32
[error]         at scala.collection.mutable.LongMap.seekEntryOrOpen(LongMap.scala:129)
[error]         at scala.collection.mutable.LongMap.getOrElseUpdate(LongMap.scala:185)
[error]         at parsley.instructions.JumpTable.apply(OptInstrs.scala:280)
[error]         at parsley.instructions.Context.runParser(Context.scala:85)
```

This is coming from the JumpTable instruction:

```
override def apply(ctx: Context): Unit =
    {
        if (ctx.moreInput)
        {
            val dest = jumpTable.getOrElseUpdate(ctx.nextChar, default) --Whoops!
            ctx.pc = dest
            if (dest == default) addErrors(ctx)
            else
            {
                ctx.checkStack = push(ctx.checkStack, ctx.offset)
                ctx.handlers = push(ctx.handlers, new Handler(ctx.depth, defaultPreamble, ctx.stack.usize))
            }
        }
        else
        {
            addErrors(ctx)
            ctx.pc = default
        }
    }
```
The getOrElseUpdate is clearly not thread safe, and so is causing a race condition with updating the table when branches are taken. I'm hoping that synchronising access to this map will fix it: it is the only piece of state left without the `Stateful` trait. Alternatively I could mark the instruction as `Stateful` and this will ensure it is copied for each thread. I'll probably do this as a hot fix, as switching to a properly thread safe version would be inefficient... Hopefully this is the _only_ issue.
